lib/chef/resource/dnf_package.rb

```diff
@@ -29,11 +29,16 @@ class DnfPackage < Chef::Resource::Package
 
       allowed_actions :install, :upgrade, :remove, :purge, :reconfig, :lock, :unlock, :flush_cache
 
-      provides :package, platform_family: %w{fedora amazon} do
-        which("dnf") && shell_out("rpm -q dnf").stdout =~ /^dnf-[1-9]/
-      end
+      # all rhel variants >= 8 will use DNF
+      provides :package, platform_family: "rhel", platform_version: ">= 8"
+
+      # fedora >= 22 uses DNF
+      provides :package, platform: "fedora", platform_version: ">= 22"
 
-      provides :package, platform_family: %{rhel}, platform_version: ">= 8"
+      # amazon will eventually use DNF
+      provides :package, platform: "amazon" do
+        which("dnf")
```

Comment:
provides gets resolved for every single resource... are we walking the filesystem looking for DNF on every single resource (and were we forking rpm each time)? This seems... awful (though not affected by this PR just a thing I noticed)...

Reply:
which just does some `File.exist?` aka `stat(2)` calls searching the `ENV['PATH']` in pure ruby, its pretty lightweight.

straight from your graph profiles:

```
   3.45%   0.01%      4.948      0.010      0.000      4.938             4265     Chef::NodeMap#block_matches?
                      4.830      0.003      0.000      4.827          124/525     Chef::Mixin::ShellOut#shell_out
                      0.055      0.004      0.000      0.051          387/387     <Class::Chef::Platform::ServiceHelpers>#service_resource_providers
                      0.046      0.002      0.000      0.044          124/254     Chef::Mixin::Which#which
                      0.005      0.005      0.000      0.000      3630/208599     NilClass#nil?
                      0.001      0.001      0.000      0.000       635/680809     Kernel#nil?
                      0.001      0.001      0.000      0.000        387/43146     Array#include?
```

So 4.83 seconds (probably about half that because of profiling overhead) in shell_out with the old code, while 0.046 seconds in `which` so almost exactly 100x faster.

And hopefully the fact that this is amazon-linux-only means it doesn't run on anything else (i would certainly *not* bet on that -- but we're now talking about tens of milliseconds total if i'm wrong and not seconds). 

That shell_out (along with the fact that the resource provides statements and the provider provides statements have drifted) is the whole reason for this.   And this should shave about 1.7 secs off your runs.

Reply:
That's on your non-IO-loaded system. I"m hacking RPM and other utilities left-and-right to not touch the disk if they don't need to...  so saving a few stats for **every single** package_resources makes a big difference.

Reply:
(those profiles are also on non-IO-loaded systems. We can get you one on a heavily IO loaded hadoop box if you wanna wait like...  a few days or so for it to run)

Reply:
Since this is `platform: "amazon"` though there's no reason for it to be running on centos.  That would be a bug that we need to fix in the NodeMap stuff.  And this is intended to be temporary and replaced once amazon finally goes DNF at some point in the future.